src/core_modules/capture-core/components/WidgetRelatedStages/hooks/useRelatedStageEligibility.js

```diff
@@ -0,0 +1,15 @@
+import { useEffect, useState } from 'react';
+
+export const useRelatedStageEligibility = (programStage, existingRelatedEvents) => {
+    const [isRelatedStageEligible, setIsRelatedStageEligible] = useState(false);
+
+    useEffect(() => {
+        if (programStage && existingRelatedEvents) {
+            const eligible = programStage.repeatable ||
+                (!programStage.repeatable && existingRelatedEvents.length === 0);
+            setIsRelatedStageEligible(eligible);
+        }
+    }, [programStage, existingRelatedEvents]);
```

Comment:
Perhaps we can refactor this to use `useMemo` instead of `useState + useEffect`. This often leads to the same effect, but with less code and performance overhead 👍 Let me know if it's unclear.